### DISCUSSION TITLE
What are the requirements for managing context in specifications?

### DISCUSSION THREAD

**Discussion by frankkilcommins** [OAI/Arazzo-Specification#7] [Ideas]
The general thinking process around the formation of this SIG, is that we want to be able to define additional context which convey the goals, or key functions, of an API (possibly even a group of APIs).

**Examples**
- The **Petstore** API has an example goal of **purchasing a pet**
- The **Auth** API has an example goal of **obtaining a refresh token**

In this discussion thread, we should aim to discuss valid use-cases and/or requirements for defining and managing the context representation. This can lead us to taking an initial pass as it's inclusion in the SIG proposal document.

Primary discussion points:
- Discuss standard for defining the key context (possible extensions to OAS, alternative specifications, etc.) with the aim of avoiding alphabetical listing of all spec items and exceed what is possible by tagging and verbose text and/or examples prose.
- Discuss wordings/terminology (e.g. recipe, goal, step, flow, workflow) - strive to avoid existing nomenclature

**Comment by ankrim** [comment]
I’m trying to categorize the different use cases to maybe help us clarify some of the details and the scope of workflows that we can or should support in this spec.
There are probably many ways to define these categories but I feel like one of the main categories is the number of “participants” in the workflow.

This comment talks about one of the OpenID Connect flows - https://github.com/OAI/sig-workflows/discussions/10#discussioncomment-1667318 and as you can see it only works with multiple participants.
On the other hand the “purchasing a pet” workflow is for a single participant.
Another example for a use case that has multiple participants is an approval flow.

I wonder if another dimension for this categorization can be single or multiple APIs, but it’s probably not as important and most of the spec should be the same in both cases.
As an example - the “purchasing a pet” workflow uses a single API (the petstore), but maybe the travel use case that was mentioned before uses multiple APIs (maybe book a flight, book an hotel room and a car rental APIs from different vendors).

I’m mostly thinking about this from a code generation perspective - how can we use the spec to generate a client to execute the flow, and in this case the flows with multiple participants are probably the most complicated. Maybe from documentation and other perspectives it’s simpler.

**Comment by orubel** [comment]
If it supports ANY KIND OF REDIRECTION, then it needs to support ALL KINDS OF REDIRECTION.

Otherwise it is only a toy for single threaded implementations.

Let me explain:

Siongle threaded applications generally do not have a front controller internally; this is placed at the app server, As such, single threaded api applications cannot do internal redirection in the app; they have to issues an EXTERNAL redirect whenever calling an api from another api. An External redirect will:
- drop the thread
- go ouytside the DMZ
- connect to DNS/router/appserver and come back in (thus doing a round robin)
- create a new thread
- resissue the request/response

This is not only VASTLY slower and more resource intensive but also potentially insecure

Multithreaded application put the front controller in the application however and thus are capable of doing internal redirection (this is talked about in Roy Fieldings dissertation BTW). This makes it so you can call an api from an api using the same request/response

This is NOT UNSAFE as you use the same JWT token to determine what ROLES and authorization you have without having to regenerate it with a secondary request/response.

This is also the REQUIREMENT for 'api chaining(r)'  https://beapi-io.github.io/beapi-docs/0.8.6/advanced.html#section-3
- API call MUST use ONLY ONE REQUEST/RESPONSE (ie no redirects)
- Client can create chains and they MUST be processed dynamically; there SHOULD BE NO hardcoded chains.
- There MUST BE no more than ONE non-safe REQUEST method used in a chain during a single REQUEST/RESPONSE
